core/src/main/java/org/axonframework/queryhandling/SimpleQueryBus.java

```diff
@@ -271,56 +247,82 @@ private boolean unsubscribe(String queryName,
             monitorCallback.reportSuccess();
         } catch (Exception e) {
             monitorCallback.reportFailure(e);
-            updateHandler.onError(e);
+            updateHandler.onCompletedExceptionally(e);
+        } finally {
+            rwLock.writeLock().unlock();
         }
 
-        return registration;
-    }
+        subscriptionQueriesRegistering.remove(interceptedQuery);
 
-    /**
-     * Invokes regular query handler and completes the update handler right afterwards.
-     *
-     * @param uow           the Unit of Work in which the query handler will be invoked
-     * @param queryHandler  the query handler to be invoked
-     * @param updateHandler the update handler to be invoked with result of query handler
-     * @param <Q>           the query type
-     * @param <I>           the initial result type
-     * @throws Exception propagated from query handler
-     */
-    private <Q, I> void invokeRegularQueryHandler(UnitOfWork<QueryMessage<Q, I>> uow,
-                                                  MessageHandler<? super QueryMessage<?, I>> queryHandler,
-                                                  UpdateHandler<I, ?> updateHandler) throws Exception {
-        I initialResult = interceptAndInvoke(uow, queryHandler).getPayload();
-        updateHandler.onInitialResult(initialResult);
-        updateHandler.onCompleted();
-    }
-
-    /**
-     * Invokes subscription query handler with freshly initialized emitter.
-     *
-     * @param uow           the Unit of Work in which the query handler will be invoked
-     * @param queryHandler  the query handler to be invoked
-     * @param updateHandler the update handler to be invoked with result of query handler
-     * @param <Q>           the query type
-     * @param <I>           the initial result type
-     * @param <U>           the incremental update type
-     * @return handle to cancel updates on this query
-     *
-     * @throws Exception propagated from query handler
-     */
-    private <Q, I, U> Registration invokeSubscriptionQueryHandler(UnitOfWork<QueryMessage<Q, I>> uow,
-                                                                  SubscriptionQueryMessageHandler<? super QueryMessage<?, I>, I, U> queryHandler,
-                                                                  UpdateHandler<I, U> updateHandler)
-            throws Exception {
-        SimpleQueryUpdateEmitter<I, U> emitter = new SimpleQueryUpdateEmitter<>(updateHandler);
-        I initialResult = interceptAndInvoke(uow, m -> queryHandler.handle(m, emitter)).getPayload();
-        emitter.initialResult(initialResult);
         return () -> {
-            emitter.cancelRegistration();
+            updateHandlers.remove(query);
             return true;
         };
     }
 
+    @SuppressWarnings("unchecked")
+    @Override
+    public <U> void emit(Predicate<SubscriptionQueryMessage<?, ?, U>> filter,
```

Comment:
Just to refresh my mind after being absent for so long; who's gonna call the `emit`, `complete` and `completeExceptionally` functions?

Reply:
Usually, event handlers, when they get an update regarding some topic. But it's not restricted to them. Complete response: any component which has some info to emit to any component that is interested to listen for it.

Reply:
Sweet, thanks for that elaboration.